README.md
# **COVID-19 ETL Pipeline**

# Preface
This README documents the purpose of the project, how to setup the automated pipeline, and future plan for development. This project was expanded based on my [COVID-19 Data Exploration Project](https://github.com/AspiringDSer/COVID19-SQL-Data-Exploration-Project) and [COVID-19 Tableau Dashboard Project](https://github.com/AspiringDSer/COVID19_Tableau_Dashboard). For more details about the initial work, please refer to the provided links.  

# Table of Content 
1. [Setup](#Setup)
2. [About the Project](#AboutTheProject)
3. [Future Developments](#FutureDevelopments)

# Setup <a name='Setup'></a>
**Environment**

In order to setup, ensure that you have `conda` installed. 

1. Open `Anaconda Prompt` and ensure you are in the correct system path 
2. Create conda environment

    ```
    conda env create -f environment.yml
    ```
3. Activate conda environment 
    ```
    conda activate COVID_ETL
    ```

**Run Pipeline**

Please execute the following commands to:

Setup ETL Pipeline 
```
python ETL_setup.py
```
Update ETL Pipeline
```
python ETL_update.py
``` 

# About The Project <a name='AboutTheProject'></a>

**ETL Pipeline**

The pipeline is used to extract and preprocess COVID-19 Data from ourworldindata.org. The cleaned data will be loaded into a MySQL Database. The database will then be connected to a Tableau Dashboard that show the total confirmed COVID-19 Cases Per Continent. 

![Confirmed Cases By Continent](./images/Confirmed_Cases_By_Continent.png)

[Tableau Public](https://public.tableau.com/views/COVID-19ConfirmedCasesByContinent/Dashboard1?:language=en-US&publish=yes&:display_count=n&:origin=viz_share_link) - Due to Tableau Public, the data is static and you cannot view the animations. The data is relevant up till Aug 2nd 2021. I have attached a [MP4 recording](https://github.com/AspiringDSer/COVID19-ETL-Pipeline/blob/master/Confirmed_Cases_By_Continent%20-%20Dashboard%20Video.mp4) (click view raw) of the animations. In addition, you can activate animations in the Tableau Workbook to see the confirmed cases per day per continent. 

**Files:**

* ETL_setup.py
* ETL_update.py
* COVID-19 Dashboard.twb
* CHANGELOG&#46;md
* environment.yml 
* Confirmed_Cases_By_Continent - Dashboard Video.mp4

`ETL_setup.py` - Initial Setup - Section 1 
1. Check if directory has existing history log 
2. Pull COVID-19 Data (CSV file) from ourworldindata.org
3. Clean data 
4. Connect to MySQL
5. Import cleaned CSV to MySQL
6. Query to create Table for Tableau Dashboard
7. Create history log

`ETL_update.py` - Update ETL Pipeline - Section 2
1. Convert history log to dataframe
2. Check if files are updated via history log
3. If data is up to date, return "Data is Up to date" message
4. If data is NOT up to date, repeat inital steps 2-7

MySQL to Tableau - Section 3 
1. Connect MySQL with Tableau 
2. Create Tableau Public Dashboard  

# Future Developments <a name='FutureDevelopments'></a>
The COVID-19 ETL Pipeline still needs further refinement. I  plan continue development in these areas: 

* Update Markdown and Comments 
* Add Unit Testing
* Improve Assertion in Python Scripts
* Configure and Implement Tableau Python (TabPy) to Pipeline
* Creating Package for Pipeline 


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html). 

## [0.1.4] - 2021-08-21
### Changed
* Updated README&#46;md

## [0.1.3] - 2021-08-05
### Changed
* Updated README&#46;md

## [0.1.2] - 2021-08-04 

### Added 
* Confirmed_Cases_By_Continent - Dashboard Video.mp4

### Changed
* Updated README&#46;md

## [0.1.1] - 2021-08-03

### Changed

* Updated .gitignore
* renamed enviornment_guide.PNG to enviornment_guide.png

## [0.1.0] - 2021-08-03

### Added 

* ETL_setup.py
* ETL_update.py
* COVID-19 Dashboard.twb
* CHANGELOG&#46;md
* environment.yml
* images/Confirmed_Cases_By_Continent.png
* images/environment_guide.png

ETL_update.py
### This script is used to check if the COVID-19 Data is up to date. ###
### If updated -> Return message that data is updated.               ###
### If NOT updated -> Repeat steps in "ETL_setup.py"                 ###

# Import Necessary Libraries 
from datetime import date
import logging 
import requests
import pandas as pd 
import sys
from ETL_setup import clean_data, connect_to_MySQL,import_csv_to_MySQL, query_table_for_tableau

def main():

    output_file_name = 'owid-covid-data.csv'
    cleaned_csv = 'cleaned_covid_data.csv'
    file_history_df = convert_filelog_to_df()

    convert_filelog_to_df()
    check_files_updated(file_history_df)
    clean_data(output_file_name)
    connect_to_MySQL()
    import_csv_to_MySQL(cleaned_csv)
    query_table_for_tableau()
    create_history_log()


# Check if files are up to date 
# Convert file_history.log to dataframe 

def convert_filelog_to_df(file_history = 'file_history.log'):

    # load in file 
    file = open(file_history, 'r')
    temp = file.read()

    # split on line space
    history = temp.split('\n')

    # create dataframe
    log_history_df = pd.DataFrame()
    log_history_df['Log History'] = history

    # removes empty last row 
    log_history_df = log_history_df[:-1]

    #create empty lists
    log_status = []
    log_date = []
    log_time = []

    # loop though each element in each row 
    for element in log_history_df['Log History']:
    
        # get log status
        log_status.append(element.split(' on')[0].split(':')[1])
    
        # get log date
        log_date.append(element.split('on ')[1].split(' ')[0])

        # get log time 
        log_time.append(element.split('on ')[1].split(' ')[1])
    
    # update dataframe  
    log_history_df['Log Status'] = log_status
    log_history_df['Log Date'] = log_date
    log_history_df['Log Time'] = log_time
    log_history_df = log_history_df.drop('Log History', axis=1)
    
    return log_history_df

# Check if files are updated by comparing the last pull timestamp with current date 
# Files are updated daily. 
def check_files_updated(history_log_df):

    # Check if Files are already updated, files are updated daily 

    today = date.today()
    check_date = today.strftime('%Y-%m-%d')

    if check_date == history_log_df['Log Date'].iloc[-1]:
        print('COVID-19 data is already up to date.')
        print('Please wait for tomorrow for updates.')
        sys.exit()
        

    else:
        print('\tPulling Updated COVID-19 DATA (CSV file)')
        with requests.get('https://covid.ourworldindata.org/data/owid-covid-data.csv') as rq:
            with open('owid-covid-data.csv', 'wb') as file:
                file.write(rq.content)

def create_history_log():
    
    # Configure Logging 
    logging.basicConfig(filename='file_history.log', level=logging.INFO,
                        format='%(levelname)s:%(message)s on %(asctime)s')

    # Create History Log
    logging.info('Successfully Updated COVID-19 Data')
    print('\tSuccessfully Updated COVID-19 Data')

if __name__ == '__main__':
    main()

ETL_setup_messy.py
# ETL Pipeline

# The Pipeline is used to extract COVID-19 Data from ourworldindata.org. 
# We are only interested in COVID-19 Cases and Deaths and the respective locations. 

## Initial Setup - Section 1 ##
# 1. Pull COVID-19 Data (CSV file) from ourworldindata.org
# 2. Pulled CSV file will be in the same directory as the script  
# 3. Create History Log
# 4. Clean the Data
# 5. Import Cleaned CSV to MySQL

## Update Data - Section 2 ##
# - Check if data is already if updated, if not, update
# - Update history log 
# - Repeat steps from Import CSV to MySQL

## MySQL to Tableau - Section 3 ##
# Connect MySQL with Tableau 
# Create Tableau Public Dashboard  



# This Python Script will be used to run Section 1 of the ETL Pipeline

# Initial Setup - Section 1 
# Import necessary libraries
import logging 
import requests
import pandas as pd 
import numpy as np
import mysql.connector
from sqlalchemy import create_engine
import sys
from clean_data import clean_data

# Configure Logging 
logging.basicConfig(filename='file_history.log', level=logging.INFO,
                    format='%(levelname)s:%(message)s on %(asctime)s')


### Pull COVID-19 Data (CSV file) from ourworldindata.org ###
### Pulled CSV file will be in the same directory as the script ###

print('\tPulling COVID-19 DATA (CSV file)')
with requests.get('https://covid.ourworldindata.org/data/owid-covid-data.csv') as rq:
    with open('owid-covid-data.csv', 'wb') as file:
        file.write(rq.content)
        
        # Create History Log
        logging.info('First Extract of COVID-19 Data')

csv = 'owid-covid-data.csv'
clean_data(csv)

# ### Clean the Data###
# print('\tCleaning Data')

# # create dataframe for csv file 
# df = pd.read_csv('owid-covid-data.csv')

# # convert `date` into a datetime data type 
# df['date'] = pd.to_datetime(df['date'])

# # Replacing all df['new_cases'] NaN values with 0 
# df['new_cases'] = df['new_cases'].fillna(0)

# # Replacing all df['new_deaths'] NaN values with 0 
# df['new_deaths'] = df['new_deaths'].fillna(0)

# # Replacing all location named Oceania to Australia
# df['location'] = df['location'].apply(lambda x: x.replace('Oceania', 'Australia'))

# # Filter out necessary columns 
# df = df[['iso_code', 'continent', 'location', 'date', 'population', 'total_cases', 'new_cases', 'total_deaths', 'new_deaths']]

# # Convert df into csv file
# df.to_csv('cleaned_covid_data.csv', header=True, index=False)

### Import Cleaned CSV to MySQL###

print('\tConnecting to MySQL')
# Connecting to MySQL
username = input('\tPlease enter MySQL username: ')
password = input('\tPlease enter MYSQL password: ')

try:
    db = mysql.connector.connect(
    user=username,
    passwd=password
)
except mysql.connector.errors.ProgrammingError:
    print('\t\tIncorrect username or password')
    print('\t\tPlease rerun script')
    sys.exit()


mycursor = db.cursor()

# Create Database
try:
    mycursor.execute("CREATE DATABASE COVID_ETL_PIPELINE")
except mysql.connector.errors.DatabaseError:
    print('\t\tCOVID_ETL_PIPELINE database already exists')
    print('\t\tWill be overwriting existing covid_19_data table')
    mycursor.execute('USE COVID_ETL_PIPELINE')
    mycursor.execute('DROP TABLE IF EXISTS covid_19_data')
 
# SQLalchemy connection to MySQL
engine = create_engine('mysql+pymysql://root:rootroot@localhost:3306/COVID_ETL_PIPELINE')

# Import Cleaned CSV to MySQL
# Make sure you are in the correct directory when running the script 

print('\tImporting Cleaned CSV to MySQL')
df = pd.read_csv("cleaned_covid_data.csv") 
df.to_sql('covid_19_data',con=engine,index=False, if_exists='append')
print('\tImport Complete')


ETL_setup.py
import logging 
import requests
import pandas as pd 
import mysql.connector
from sqlalchemy import create_engine
import sys
import os

def main():

    url = 'https://covid.ourworldindata.org/data/owid-covid-data.csv'
    output_file_name = 'owid-covid-data.csv'
    cleaned_csv = 'cleaned_covid_data.csv'

    check_if_history_file_exists()
    pull_data(url, output_file_name)
    clean_data(output_file_name)
    connect_to_MySQL()
    import_csv_to_MySQL(cleaned_csv)
    query_table_for_tableau()
    create_history_log()

def check_if_history_file_exists():
    
    print('\tChecking if file history log exists in this directory')

    if os.path.exists('file_history.log'):
        print('\t\tA History log exists.')
        print('\t\tWould you like to delete it? (yes or no)')
        
        while True:

            answer = input('\t\tyes or no: ')
            if answer == 'yes':
                os.remove('file_history.log')
                print('\t\tHistory log removed.')
                return False

            elif answer == 'no':
                print('\t\tPlease rename the existing log to something other than "file_history.log".')
                print('\t\tPlease rerun the script after completing the rename.')
                sys.exit()

            else:
                print('\t\tIncorrect answer. Please choose "yes" or "no"')           

def pull_data(url, output_file_name):

    print('\tPulling COVID-19 DATA (CSV file)')
    with requests.get(url) as rq:
        with open(output_file_name, 'wb') as file:
            file.write(rq.content)

def clean_data(output_file_name):
    
    print('\tCleaning Data')

    # create dataframe for csv file 
    df = pd.read_csv(output_file_name)

    # convert `date` into a datetime data type 
    df['date'] = pd.to_datetime(df['date'])

    # Replacing all df['new_cases'] NaN values with 0 
    df['new_cases'] = df['new_cases'].fillna(0)

    # Replacing all df['new_deaths'] NaN values with 0 
    df['new_deaths'] = df['new_deaths'].fillna(0)

    # Replacing all location named Oceania to Australia
    df['location'] = df['location'].apply(lambda x: x.replace('Oceania', 'Australia'))

    # Filter out necessary columns 
    df = df[['iso_code', 'continent', 'location', 'date', 'population', 'total_cases', 'new_cases', 'total_deaths', 'new_deaths']]

    # Convert df into csv file
    df.to_csv('cleaned_covid_data.csv', header=True, index=False)

def connect_to_MySQL():
    
    print('\tConnecting to MySQL')
    # Connecting to MySQL
    try:

        global username 
        username = input('\tPlease enter MySQL username: ')
    
        global password 
        password = input('\tPlease enter MYSQL password: ')
    
        db = mysql.connector.connect(
             user=username,
             passwd=password)

    except mysql.connector.errors.ProgrammingError:
        print('\t\tIncorrect username or password')
        print('\t\tPlease rerun script')
        sys.exit()

    mycursor = db.cursor()

    # Create Database
    try:
        mycursor.execute("CREATE DATABASE COVID_ETL_PIPELINE")
    except mysql.connector.errors.DatabaseError:
        print('\t\tCOVID_ETL_PIPELINE database already exists')
        print('\t\tWill be overwriting existing covid_19_data table')
        mycursor.execute('USE COVID_ETL_PIPELINE')
        mycursor.execute('DROP TABLE IF EXISTS covid_19_data')

def import_csv_to_MySQL(cleaned_csv):
    # Import Cleaned CSV to MySQL
    # Make sure you are in the correct directory when running the script 

    # SQLalchemy connection to MySQL
    engine = create_engine(f'mysql+pymysql://{username}:{password}@localhost:3306/COVID_ETL_PIPELINE') 

    print('\tImporting Cleaned CSV to MySQL')
    df = pd.read_csv(cleaned_csv) 
    df.to_sql('covid_19_data',con=engine,index=False, if_exists='append')
    print('\tImport Completed')

def query_table_for_tableau():
    # Create table needed for Tableau Dashboard 

    print('\tCreating Table for Tableau Dashboard')
    db = mysql.connector.connect(
         user=username,
         passwd=password)

    mycursor = db.cursor()

    mycursor.execute('USE covid_etl_pipeline')
    mycursor.execute('DROP TABLE IF EXISTS Total_Cases_Per_Continent')
    mycursor.execute('CREATE TABLE Total_Cases_Per_Continent (date date,location varchar(255),cumulative_cases numeric)') 
    mycursor.execute('INSERT INTO Total_Cases_Per_Continent SELECT date, location, SUM(new_cases) OVER w AS cumulative_cases FROM covid_19_data WHERE location IN ("Europe", "South America", "North America", "Africa", "Australia", "Asia") WINDOW w AS (PARTITION BY location ORDER BY date)')       
    db.commit()                    

def create_history_log():
    
    # Configure Logging 
    logging.basicConfig(filename='file_history.log', level=logging.INFO,
                        format='%(levelname)s:%(message)s on %(asctime)s')

    # Create History Log
    logging.info('First Extract of COVID-19 Data')
    print('\tInitial History Log Created')

if __name__ == '__main__':
    main()